src/gleam/list.gleam

```diff
@@ -1452,3 +1452,49 @@ pub fn last(list: List(a)) -> Result(a, Nil) {
   list
   |> reduce(fn(elem, _) { elem })
 }
+
+/// Return unique combinations of elements in the list
+///
+/// ## Examples
+///
+/// ```
+/// > combinations_by([1, 2, 3], 2)
+/// [[1, 2], [1, 3], [2, 3]]
+///
+///  > combinations_by([1, 2, 3, 4], 3)
+///  [[1, 2, 3], [1, 2, 4], [1, 3, 4], [2, 3, 4]]
+/// ```
+///
+pub fn combinations_by(items: List(a), n: Int) -> List(List(a)) {
+  case n {
+    0 -> [[]]
+    _ ->
+      case items {
+        [] -> []
+        [x, ..xs] -> {
+          let first_combinations =
+            map(combinations_by(xs, n - 1), with: fn(com) { [x, ..com] })
+          append(first_combinations, combinations_by(xs, n))
```

Comment:
Unfortunately the append function is quite expensive and not ideal for using in a loop. Could we build a list of lists and then flatten it at the end? That would save a lot of repeated copying

Reply:
Looking into this

Reply:
Looking at the implementation of `flatten`, that function still calls `append` for each element in the list. So it seems using flatten won't make a difference?

Reply:
If we use flattern append is called O(n) times where n the the length of the list, while with append in the loop it's O(n log n) times

Reply:
I can't figure out how to do this in two steps here.
But I can use `fold_right` instead of `append`

Instead of 
`append(first_combinations, combinations(xs, n))
`
Do you think this would be any faster?

```
fold_right(
  first_combinations,
  combinations(xs, n),
  fn(c, acc) { [c, ..acc] },
)
```

I guess `[c, ..acc]` should be cheap, right?

Reply:
Or using reverse and fold

```
          let first_combinations =
            map(combinations(xs, n - 1), with: fn(com) { [x, ..com] })
            |> reverse
          fold(
            first_combinations,
            combinations(xs, n),
            fn(c, acc) { [c, ..acc] },
          )
```

Reply:
Is the ordering important here?

Reply:
Not really, you get the same combinations in different order. We can drop the `reverse`

Reply:
With the `'reverse' the return is a bit more intuitive

e.g.
```
[[1,2],[1,3],[2,3]]
```

You see 1,2 then 1,3

Without reverse

```
[[1,3],[1,2],[2,3]]
```
But the combinations are the same